proposals/007-remove-imports-from-components.md

```diff
@@ -0,0 +1,52 @@
+# Remove the imports of settings, tools and helpers layers from components
+
+**Closing date:** 27 March 2020
+
+**Status:** Proposed
+
+## Summary
+
+We plan to remove the imports of the settings, helpers and tools layers from all components in GOV.UK Frontend version 4.0.
+
+We’re doing this to reduce the time it takes to compile GOV.UK Frontend’s Sass to CSS, especially if you’re using the Ruby Sass compiler.
+
+## Problem
+
+If you're using the Ruby Sass compiler to compile Sass to CSS, it can take a long time because GOV.UK Frontend has a lot of imports.
+
+Although Ruby Sass is deprecated, teams (including GOV.UK) are stuck using it for the foreseeable future.
```

Comment:
It would be good to have background on why teams are stuck with Ruby Sass? Looking at the benchmarks below, it doesn't seem like huge issue with the other compilers - i.e. 0.02s, 0.25s less seems hardly noticeable. Should we be focusing on moving away from Ruby Sass instead?

Reply:
There is a big problem with us (GOV.UK) switching to libsass with govuk-frontend having the large number of imports. This reduces the time of initial compile by about 50% but then has a massively slow time for repeated requests (in [Finder Frontend on govuk-docker](https://github.com/alphagov/govuk-frontend/issues/1671#issue-532246501) a warm libsass cache took 17s responses compared to 0.22s with RubySass). I'm pretty sure this is all a problem with Sprockets' usage of libsass rather than any issues in libsass and it's rather unfortunate as it's a real bummer still using RubySass.

I don't think it's quite a forseeable future issue though I think within a few months if we can dedicate some resources it's resolvable (depending on the current crisis) there has been talking about putting together a frontend dev based group together to regularly catch up on this.

Aside from the big warm cache issue (which would be mostly resolved by this proposal) the other issues we've got are:

- [Rails SASS compressor breaking quoted SCSS functions](https://github.com/alphagov/finder-frontend/issues/1124) - which we might [have a fix](https://github.com/alphagov/whitehall/commit/1ae8699230617783533979d77b9d71a54198b752) for
- Switching to a non Ruby Sass based linter, currently the linter gem actually makes most GOV.UK apps use RubySass by depending on the deprecated gem which Sprockets notices and uses.